core/builder/src/main/java/io/quarkus/builder/BuildChainBuilder.java

```diff
@@ -417,19 +425,46 @@ private StepInfo buildOne(BuildStepBuilder toBuild, Set<BuildStepBuilder> includ
                 includedDependencies++;
             }
         }
-        int includedDependents = 0;
-        for (BuildStepBuilder dependent : dependentsOfThis) {
-            if (included.contains(dependent)) {
-                includedDependents++;
+        toBuild.dependencyCount(includedDependencies);
+
+        if (toBuild.getDependencyCount() == 0) {
+            // it's a start step!
+            startSteps.add(toBuild);
+        }
+
+    }
+
+    void setOrdinals(Set<BuildStepBuilder> startSteps) {
```

Comment:
Could the ordinals now be set at the time the build steps are each actually added?

Reply:
Yes, they can, thanks for asking :) Calling `buildOne()` for each `included` is actually a depth-first search - a well known topological sort algorithm.

Reply:
Solved in 351945e